File: src/pages/login.page.ts
import { Page } from "playwright";
import { waitForVisible } from "../utils/wait-utils";

export class LoginPage {
  private readonly page: Page;

  // TODO: Replace with app-specific locators.
  private readonly usernameSelector = "#username";
  private readonly passwordSelector = "#password";
  private readonly loginButtonSelector = "#loginBtn";
  private readonly dashboardSelector = "[data-test='dashboard']";
  private readonly loginErrorSelector = "[data-test='login-error']";

  constructor(page: Page) {
    this.page = page;
  }

  async open(url: string): Promise<void> {
    await this.page.goto(url, { waitUntil: "domcontentloaded" });
  }

  async enterUsername(username: string): Promise<void> {
    const usernameInput = this.page.locator(this.usernameSelector);
    await waitForVisible(usernameInput);
    await usernameInput.fill(username);
  }

  async enterPassword(password: string): Promise<void> {
    const passwordInput = this.page.locator(this.passwordSelector);
    await waitForVisible(passwordInput);
    await passwordInput.fill(password);
  }

  async clickLogin(): Promise<void> {
    const loginButton = this.page.locator(this.loginButtonSelector);
    await waitForVisible(loginButton);
    await loginButton.click();
  }

  async isDashboardVisible(): Promise<boolean> {
    const dashboardMarker = this.page.locator(this.dashboardSelector);
    await waitForVisible(dashboardMarker);
    return dashboardMarker.isVisible();
  }

  async isLoginErrorVisible(): Promise<boolean> {
    const loginError = this.page.locator(this.loginErrorSelector);
    await waitForVisible(loginError);
    return loginError.isVisible();
  }
}

File: src/test/java/com/automation/pages/LoginPage.java
package com.automation.pages;

import com.automation.utils.WaitUtils;
import org.openqa.selenium.By;
import org.openqa.selenium.WebDriver;

public class LoginPage {

    private final WebDriver driver;

    // TODO: Update locators based on actual application DOM.
    private final By usernameInput = By.id("username");
    private final By passwordInput = By.id("password");
    private final By loginButton = By.id("loginBtn");
    private final By dashboardMarker = By.cssSelector("[data-test='dashboard']");
    private final By loginError = By.cssSelector("[data-test='login-error']");

    public LoginPage(WebDriver driver) {
        this.driver = driver;
    }

    public void open(String loginUrl) {
        driver.get(loginUrl);
    }

    public void enterUsername(String username) {
        WaitUtils.waitForVisible(driver, usernameInput).sendKeys(username);
    }

    public void enterPassword(String password) {
        WaitUtils.waitForVisible(driver, passwordInput).sendKeys(password);
    }

    public void clickLogin() {
        WaitUtils.waitForClickable(driver, loginButton).click();
    }

    public boolean isDashboardVisible() {
        return WaitUtils.waitForVisible(driver, dashboardMarker).isDisplayed();
    }

    public boolean isLoginErrorVisible() {
        return WaitUtils.waitForVisible(driver, loginError).isDisplayed();
    }
}
